src/index.ts

```diff
@@ -40,7 +40,7 @@ export function decode<Output, Input>(
       return fold<t.Errors, Output, Promise<Output>>(
         errors => Promise.reject(new DecodeError(errors)),
         decodedValue => Promise.resolve(decodedValue),
-      )(type.decode(value || arguments[2]));
```

Comment:
Hmm, the code here looks a bit strange. The `arguments` list should never have a third element, so the reference to `arguments[2]` doesn't make any sense. Basically `value` should always be passed to the decoder, but TypeScript will complain about it being possibly `undefined`, so that is probably why I defaulted to the arguments list, but was off by one.

I think a better fix for the issue would be to fix the arguments list index. This way `undefined` values will also decode properly.

Reply:
Hm, yes took me a while before I figured out what the `arguments[2]` part was about. I agree that there should be a better fix for this but I'm quite new to TypeScript and can't seem to figure out how to convince it to accept `undefined` 😊.

Reply:
Okay, no worries. So the need for the `|| arguments [x]` comes from the fact that TypeScript can't determine from the `arguments.length` switch statement that the type of `value` would always be defined. The problem in the original code was of course that it indexed the wrong, nonexistent, third argument, which caused the bug you encountered.

From a runtime point of view 

```typescript
type.decode(value)
```

would work without problems, but we need to define the fallback in order to satisfy TypeScript type checks.

Since `value` and `arguments[1]` are actually the same value, the following would also work for both runtime and TypeScript since `arguments[1]` has the type `any`.

```typescript
type.decode(arguments[1])
```

But has the problem of linters being picky and complaining that `value` is never used.

So to satisfy all tools, the following is probably the best we can do

```typescript
type.decode(value || arguments[1])
```

Here `value` is used, and provides some useful type checks to ensure that the passed value is compatible with the parameter for `type.decode`. And then `|| arguments[1]` is used as a fallback for TypeScript type safety.

```typescript
value === undefined ? arguments[1] : value
```

is also completely valid solution, and might even be a bit more clear to the reader.

Reply:
Ok, thank you for the explanation! Fixed the argument index using your solution.